### PR TITLE
Make that `platform.set_parameters()` doesnt need connection

### PR DESCRIPTION
**Pull request by GuillermoAbadLopez**
Steps to be done:

- [x] Adding if self.device is not None in all sets/setups of instruments, buses, etc… 

- [x] Remove the @CheckDeviceInitialised decorator from setups

- [x] Make sure platform.initial_setup raises an error when we are not connected

- [x] Add documentation in platform.set_parameter to make sure users understand what happens if they are/aren't connected to the instruments.

- [x] Change tests to work with `if self.device is not None` instead than with the flag:

- [x] Changed `system_controll`, so the `port_id` flag is removed from everywhere else in qililab

### CODE REVIEW

**Review by AlbertMitjans — COMMENTED** [review]
Guille thank you for the effort!

In the future let's make sure we refine big tasks like this before jumping into them, to avoid losing time implementing stuff that is not optimal.

In this specific case, I don't like the `instrument_set` argument that you added, because we don't need it at all.

I believe the best solution here is to avoid doing the `self.device.set...` when `self.device` is `None`. In other words, if the user calls `set_parameter` when the instrument is not connected, only the value of the cache will be updated.

File: src/qililab/instruments/qblox/qblox_module.py [hunk 1]
```diff
@@ -608,10 +668,16 @@ def _set_gain_imbalance(self, value: float | str | bool, sequencer_id: int):
         """
 
         self._get_sequencer_by_id(id=sequencer_id).gain_imbalance = float(value)
-        self.device.sequencers[sequencer_id].mixer_corr_gain_ratio(float(value))
+        if instrument_set:
+            self.device.sequencers[sequencer_id].mixer_corr_gain_ratio(float(value))
```
Comment (AlbertMitjans) [inline src/qililab/instruments/qblox/qblox_module.py:672]:
```suggestion
        if self.device is not None:
            self.device.sequencers[sequencer_id].mixer_corr_gain_ratio(float(value))
```

Comment (GuillermoAbadLopez) [reply src/qililab/instruments/qblox/qblox_module.py]: In the end after a discussion, we decided we are going to check it with:
`if hasattr(self, "device") and self.device is not None:`

**Comment by GuillermoAbadLopez** [comment]
@AlbertMitjans can `self.device` be `None`? In the code its seems to be checked with: `if hasattr(self, "device")`, so I wrote: `if hasattr(self, "device") and self.device is not None:` everywhere, but maybe the last part is not actually needed??

File: src/qililab/instruments/agilent/e5071b_vna.py [hunk 2]
```diff
@@ -51,13 +52,15 @@ def electrical_delay(self, time: float):
             value (str) : Electrical delay in ns
         """
         self.settings.electrical_delay = time
-        self.send_command("CALC:MEAS:CORR:EDEL:TIME", f"{time}")
+        if hasattr(self, "device") and self.device is not None:
```
Comment (fedonman) [inline src/qililab/instruments/agilent/e5071b_vna.py:55]: Since this check is repeated everywhere you should create a method that includes it, for example `is_device_initialized()`. It can be in the `Instrument` class level, since `device` is common for all instruments.

File: src/qililab/instruments/mini_circuits/attenuator.py [hunk 3]
```diff
@@ -43,13 +43,19 @@ class StepAttenuatorSettings(Instrument.InstrumentSettings):
...
 
-    @Instrument.CheckDeviceInitialized
     @Instrument.CheckParameterValueFloatOrInt
-    def setup(self, parameter: Parameter, value: float | str | bool, channel_id: int | None = None):
+    def setup(  # type: ignore
+        self,
+        parameter: Parameter,
+        value: float | str | bool,
+        channel_id: int | None = None,
+        port_id: str | None = None,
```
Comment (fedonman) [inline src/qililab/instruments/mini_circuits/attenuator.py:52]: why are we adding `port_id`?

File: src/qililab/instruments/awg_analog_digital_converter.py [hunk 4]
```diff
@@ -49,9 +49,12 @@ def acquire_result(self) -> Result:
...
         """
 
-    @Instrument.CheckDeviceInitialized
     def setup(  # pylint: disable=too-many-return-statements, too-many-branches
-        self, parameter: Parameter, value: float | str | bool, channel_id: int | None = None
+        self,
+        parameter: Parameter,
+        value: float | str | bool,
+        channel_id: int | None = None,
+        port_id: str | None = None,
```
Comment (fedonman) [inline src/qililab/instruments/awg_analog_digital_converter.py:57]: Here too, why are we adding `port_id`?

File: src/qililab/instruments/instrument.py [hunk 5]
```diff
@@ -160,14 +160,19 @@ def __init__(self, settings: dict):
...
         """Set initial instrument settings."""
 
-    @CheckDeviceInitialized
-    def setup(self, parameter: Parameter, value: float | str | bool, channel_id: int | None = None):
+    def setup(
+        self,
+        parameter: Parameter,
+        value: float | str | bool,
+        channel_id: int | None = None,
+        port_id: str | None = None,
```
Comment (fedonman) [inline src/qililab/instruments/instrument.py:168]: Here too, why are we adding `port_id`?

File: src/qililab/instruments/instrument.py [hunk 6]
```diff
@@ -160,14 +160,19 @@ def __init__(self, settings: dict):
...
+    ):
         """Set instrument settings parameter to the corresponding value
 
         Args:
             parameter (Parameter): settings parameter to be updated
             value (float | str | bool): new value
-            channel_id (int | None): channel identifier of the parameter to update
+            channel_id (int | None): channel identifier of the parameter to update\
```
Comment (fedonman) [inline src/qililab/instruments/instrument.py:175]: added `\` at the end of the docstring

File: src/qililab/instruments/keithley/keithley_2600.py [hunk 7]
```diff
@@ -44,22 +44,24 @@ class Keithley2600Settings(Instrument.InstrumentSettings):
...
 
-    @Instrument.CheckDeviceInitialized
     @Instrument.CheckParameterValueFloatOrInt
-    def setup(
+    def setup(  # type: ignore
         self,
         parameter: Parameter,
         value: float | str | bool,
         channel_id: int | None = None,
+        port_id: str | None = None,
```
Comment (fedonman) [inline src/qililab/instruments/keithley/keithley_2600.py:53]: Here too, why are we adding `port_id`?

File: src/qililab/instruments/qblox/qblox_d5a.py [hunk 8]
```diff
@@ -74,8 +74,13 @@ def _channel_setup(self, dac_index: int) -> None:
...
             sleep(0.1)
 
-    @Instrument.CheckDeviceInitialized
-    def setup(self, parameter: Parameter, value: float | str | bool, channel_id: int | None = None):
+    def setup(
+        self,
+        parameter: Parameter,
+        value: float | str | bool,
+        channel_id: int | None = None,
+        port_id: str | None = None,
```
Comment (fedonman) [inline src/qililab/instruments/qblox/qblox_d5a.py:82]: Same here, why are we adding port_id?

File: src/qililab/instruments/qblox/qblox_qcm_rf.py [hunk 9]
```diff
@@ -70,9 +70,12 @@ def initial_setup(self):
...
             self.setup(parameter, getattr(self.settings, parameter.value))
 
-    @Instrument.CheckDeviceInitialized
     def setup(
-        self, parameter: Parameter, value: float | str | bool, channel_id: int | None = None, port_id: str | None = None
+        self,
+        parameter: Parameter,
+        value: float | str | bool,
+        channel_id: int | None = None,
+        port_id: str | None = None,
```
Comment (fedonman) [inline src/qililab/instruments/qblox/qblox_qcm_rf.py:78]: Same here why are we adding port_id?

File: src/qililab/instruments/qblox/qblox_qrm.py [hunk 10]
```diff
@@ -374,12 +382,21 @@ def integration_length(self, sequencer_id: int):
...
         return cast(AWGQbloxADCSequencer, self.get_sequencer(sequencer_id)).integration_length
 
-    @Instrument.CheckDeviceInitialized
     def setup(
-        self, parameter: Parameter, value: float | str | bool, channel_id: int | None = None, port_id: str | None = None
+        self,
+        parameter: Parameter,
+        value: float | str | bool,
+        channel_id: int | None = None,
+        port_id: str | None = None,
```
Comment (fedonman) [inline src/qililab/instruments/qblox/qblox_qrm.py:390]: Here too, why are we adding port_id?

File: src/qililab/instruments/qblox/qblox_qrm.py [hunk 11]
```diff
@@ -374,12 +382,21 @@ def integration_length(self, sequencer_id: int):
...
         try:
             AWGAnalogDigitalConverter.setup(self, parameter=parameter, value=value, channel_id=channel_id)
         except ParameterNotFound:
-            QbloxModule.setup(self, parameter=parameter, value=value, channel_id=channel_id, port_id=port_id)
+            QbloxModule.setup(
+                self,
+                parameter=parameter,
+                value=value,
+                channel_id=channel_id,
+                port_id=port_id,
```
Comment (fedonman) [inline src/qililab/instruments/qblox/qblox_qrm.py:401]: Same here.

File: src/qililab/instruments/qblox/qblox_qrm_rf.py [hunk 12]
```diff
@@ -60,9 +60,12 @@ def initial_setup(self):
...
             self.setup(parameter, getattr(self.settings, parameter.value))
 
-    @Instrument.CheckDeviceInitialized
     def setup(
-        self, parameter: Parameter, value: float | str | bool, channel_id: int | None = None, port_id: str | None = None
+        self,
+        parameter: Parameter,
+        value: float | str | bool,
+        channel_id: int | None = None,
+        port_id: str | None = None,
```
Comment (fedonman) [inline src/qililab/instruments/qblox/qblox_qrm_rf.py:68]: Same here.

File: src/qililab/instruments/qblox/qblox_s4g.py [hunk 13]
```diff
@@ -73,8 +73,13 @@ def _channel_setup(self, dac_index: int) -> None:
...
             sleep(0.1)
 
-    @Instrument.CheckDeviceInitialized
-    def setup(self, parameter: Parameter, value: float | str | bool, channel_id: int | None = None):
+    def setup(
+        self,
+        parameter: Parameter,
+        value: float | str | bool,
+        channel_id: int | None = None,
+        port_id: str | None = None,
```
Comment (fedonman) [inline src/qililab/instruments/qblox/qblox_s4g.py:81]: Same here.

File: src/qililab/instruments/qdevil/qdevil_qdac2.py [hunk 14]
```diff
@@ -54,8 +54,13 @@ def low_pass_filter(self):
...
         return self.settings.low_pass_filter
 
-    @Instrument.CheckDeviceInitialized
-    def setup(self, parameter: Parameter, value: float | str | bool, channel_id: int | None = None):
+    def setup(  # pylint: disable=too-many-branches
+        self,
+        parameter: Parameter,
+        value: float | str | bool,
+        channel_id: int | None = None,
+        port_id: str | None = None,
```
Comment (fedonman) [inline src/qililab/instruments/qdevil/qdevil_qdac2.py:62]: Same here.

File: src/qililab/instruments/rohde_schwarz/sgs100a.py [hunk 15]
```diff
@@ -43,31 +43,34 @@ class SGS100ASettings(SignalGenerator.SignalGeneratorSettings):
     settings: SGS100ASettings
     device: RohdeSchwarzSGS100A
 
-    @Instrument.CheckDeviceInitialized
     def setup(
         self,
         parameter: Parameter,
         value: float | str | bool,
         channel_id: int | None = None,
+        port_id: str | None = None,
```
Comment (fedonman) [inline src/qililab/instruments/rohde_schwarz/sgs100a.py:51]: Same here.

File: src/qililab/instruments/vector_network_analyzer.py [hunk 16]
```diff
@@ -61,8 +61,13 @@ class VectorNetworkAnalyzerSettings(Instrument.InstrumentSettings):  # pylint: d
...
     device: VectorNetworkAnalyzerDriver
 
-    @Instrument.CheckDeviceInitialized
-    def setup(self, parameter: Parameter, value: float | str | bool | int, channel_id: int | None = None):
+    def setup(
+        self,
+        parameter: Parameter,
+        value: float | str | bool | int,
+        channel_id: int | None = None,
+        port_id: str | None = None,
```
Comment (fedonman) [inline src/qililab/instruments/vector_network_analyzer.py:69]: Same here.

File: src/qililab/instruments/yokogawa/gs200.py [hunk 17]
```diff
@@ -151,14 +154,20 @@ def voltage(self) -> float:
...
+                self.device.ramp_voltage(value, self.ramping_rate * 0.001, 0.001)
+            else:
+                self.device.voltage(value)
+
+    def setup(
+        self,
+        parameter: Parameter,
+        value: float | str | bool,
+        channel_id: int | None = None,
+        port_id: str | None = None,
```
Comment (fedonman) [inline src/qililab/instruments/yokogawa/gs200.py:169]: Same here.

File: src/qililab/system_control/system_control.py [hunk 18]
```diff
@@ -132,7 +132,11 @@ def instruments(self) -> list[Instrument]:
         return self.settings.instruments
 
     def set_parameter(
-        self, parameter: Parameter, value: float | str | bool, channel_id: int | None = None, port_id: str | None = None
+        self,
+        parameter: Parameter,
+        value: float | str | bool,
+        channel_id: int | None = None,
+        port_id: str | None = None,
```
Comment (fedonman) [inline src/qililab/system_control/system_control.py:139]: Same here.

**Review by fedonman — CHANGES_REQUESTED** [review]
Thanks for the super quick changes! :100 

Left some comments on the code, I will check the tests too after addressing the comments.

**Comment by fedonman** [comment]
> @AlbertMitjans can `self.device` be `None`? In the code its seems to be checked with: `if hasattr(self, "device")`, so I wrote: `if hasattr(self, "device") and self.device is not None:` everywhere, but maybe the last part is not actually needed??

If you implement this review comment https://github.com/qilimanjaro-tech/qililab/pull/647#discussion_r1434872662 it will be very easy to test which if statement is the ideal, since you will only have to change one line.

`if hasattr(self, "device")` is correct for sure, I might also check if it is a subclass of `Device`: `isinstance(self.device, Device)`.

File: src/qililab/instruments/agilent/e5071b_vna.py (hunk 2 above)
Comment (GuillermoAbadLopez) [reply src/qililab/instruments/agilent/e5071b_vna.py:55]: That's a great idea!

File: src/qililab/instruments/mini_circuits/attenuator.py (hunk 3 above)
Comment (GuillermoAbadLopez) [reply src/qililab/instruments/mini_circuits/attenuator.py:52]: pylint/mypy were complaining, that the child's didn't follow the parent interfaces,

So I could add an ignore or fix it, I don't think that breaks anything, but if you think we shouldn't do it anyway let me know again!

File: src/qililab/instruments/instrument.py (hunk 5 above)
Comment (GuillermoAbadLopez) [reply src/qililab/instruments/instrument.py:168]: Same answer aboeve

File: src/qililab/instruments/awg_analog_digital_converter.py (hunk 4 above)
Comment (GuillermoAbadLopez) [reply src/qililab/instruments/awg_analog_digital_converter.py:57]: Same answer aboeve

File: src/qililab/instruments/instrument.py (hunk 6 above)
Comment (GuillermoAbadLopez) [reply src/qililab/instruments/instrument.py:175]: nice catch!

File: src/qililab/instruments/keithley/keithley_2600.py (hunk 7 above)
Comment (GuillermoAbadLopez) [reply src/qililab/instruments/keithley/keithley_2600.py:53]: Same answer aboeve

File: src/qililab/instruments/qblox/qblox_qcm_rf.py (hunk 9 above)
Comment (GuillermoAbadLopez) [reply src/qililab/instruments/qblox/qblox_qcm_rf.py:78]: Same answer aboeve

File: src/qililab/instruments/qblox/qblox_d5a.py (hunk 8 above)
Comment (GuillermoAbadLopez) [reply src/qililab/instruments/qblox/qblox_d5a.py:82]: Same answer aboeve

File: src/qililab/instruments/qblox/qblox_qrm.py (hunk 11 above)
Comment (GuillermoAbadLopez) [reply src/qililab/instruments/qblox/qblox_qrm.py:401]: Same answer aboeve

File: src/qililab/instruments/qblox/qblox_qrm_rf.py (hunk 12 above)
Comment (GuillermoAbadLopez) [reply src/qililab/instruments/qblox/qblox_qrm_rf.py:68]: Same answer aboeve

File: src/qililab/instruments/qdevil/qdevil_qdac2.py (hunk 14 above)
Comment (GuillermoAbadLopez) [reply src/qililab/instruments/qdevil/qdevil_qdac2.py:62]: Same answer aboeve

File: src/qililab/instruments/rohde_schwarz/sgs100a.py (hunk 15 above)
Comment (GuillermoAbadLopez) [reply src/qililab/instruments/rohde_schwarz/sgs100a.py:51]: Same answer aboeve

File: src/qililab/instruments/vector_network_analyzer.py (hunk 16 above)
Comment (GuillermoAbadLopez) [reply src/qililab/instruments/vector_network_analyzer.py:69]: Same answer aboeve

File: src/qililab/instruments/yokogawa/gs200.py (hunk 17 above)
Comment (GuillermoAbadLopez) [reply src/qililab/instruments/yokogawa/gs200.py:169]: Same answer aboeve

File: src/qililab/system_control/system_control.py (hunk 18 above)
Comment (GuillermoAbadLopez) [reply src/qililab/system_control/system_control.py:139]: Same answer aboeve

File: src/qililab/instruments/qblox/qblox_s4g.py (hunk 13 above)
Comment (GuillermoAbadLopez) [reply src/qililab/instruments/qblox/qblox_s4g.py:81]: Same answer aboeve

File: src/qililab/instruments/qblox/qblox_qrm.py (hunk 10 above)
Comment (GuillermoAbadLopez) [reply src/qililab/instruments/qblox/qblox_qrm.py:390]: Same answer aboeve

**Comment by GuillermoAbadLopez** [comment]
> > @AlbertMitjans can `self.device` be `None`? In the code its seems to be checked with: `if hasattr(self, "device")`, so I wrote: `if hasattr(self, "device") and self.device is not None:` everywhere, but maybe the last part is not actually needed??
> 
> If you implement this review comment [#647 (comment)](https://github.com/qilimanjaro-tech/qililab/pull/647#discussion_r1434872662) it will be very easy to test which if statement is the ideal, since you will only have to change one line.
> 
> `if hasattr(self, "device")` is correct for sure, I might also check if it is a subclass of `Device`: `isinstance(self.device, Device)`.

But you wouldn't check if it is None, then? I'll implement your suggestion and we can try it out quicker!

File: src/qililab/instruments/mini_circuits/attenuator.py (hunk 3 above)
Comment (fedonman) [reply src/qililab/instruments/mini_circuits/attenuator.py:52]: Yes, it's complaining because you have added it in `Instrument` as well. The question is why adding it in the first place.

File: src/qililab/instruments/keithley/keithley_2600.py [hunk 19]
```diff
@@ -55,12 +55,12 @@ def setup(  # type: ignore
         """Setup instrument."""
         if parameter == Parameter.MAX_CURRENT:
             self.max_current = float(value)
-            if hasattr(self, "device") and self.device is not None:
+            if Instrument.is_device_initialized(self):
```
Comment (fedonman) [inline src/qililab/instruments/keithley/keithley_2600.py:58]: Does this syntax work? Why not `self.is_device_initialized()` ?

File: src/qililab/instruments/mini_circuits/attenuator.py (hunk 3 above)
Comment (GuillermoAbadLopez) [reply src/qililab/instruments/mini_circuits/attenuator.py:52]: It was complaining before me doing any change Vyron, I didn't add more code because I felt like 😅

I just checked, and the reason was that `qblox_module`, `qblox_qrm`'s and `qblox_qcm`'s had that parameter in the `setup`, so pylint or mypy was complaining 🙃

Do you think I shouldn't expand the qblox_behaviour up to the parent then? I could place ignores in those qblox child instead than expanding it 👌

File: src/qililab/instruments/keithley/keithley_2600.py (hunk 19 above)
Comment (GuillermoAbadLopez) [reply src/qililab/instruments/keithley/keithley_2600.py:58]: Lmao, you are totally right, dunno why I did that! 

Changing it!

File: src/qililab/instruments/qblox/qblox_module.py [hunk 20]
```diff
@@ -412,10 +419,16 @@ def _set_hardware_modulation(self, value: float | str | bool, sequencer_id: int)
...
+        if Instrument.is_device_initialized(self):
+            self.device.sequencers[sequencer_id].mod_en_awg(bool(value))
 
     @Instrument.CheckParameterValueFloatOrInt
-    def _set_frequency(self, value: float | str | bool, sequencer_id: int):
+    def _set_frequency(
+        self,
+        value: float | str | bool,
+        sequencer_id: int,
+    ):
```
Comment (AlbertMitjans) [inline src/qililab/instruments/qblox/qblox_module.py:431]: Do you have the correct `line-length` setup in your VSCode ([look at the last setting of the JSON](https://www.notion.so/qilimanjaro/Code-guidelines-199b362b39af4261a2eaa512051145bb?pvs=4#fba4f5ca731d4b019e508d9134eb1c07))?

If you do, removing the comma of the last argument and saving should move the code back to a single line. 🙏

Comment (AlbertMitjans) [reply src/qililab/instruments/qblox/qblox_module.py:431]: This way the PR will have much less line changes!

Comment (GuillermoAbadLopez) [reply src/qililab/instruments/qblox/qblox_module.py:431]: Oh! I didn't know a coma made the auto-formater put it in vertical!

I'll change it, sure!

**Comment by GuillermoAbadLopez** [comment]
> > > @AlbertMitjans can `self.device` be `None`? In the code its seems to be checked with: `if hasattr(self, "device")`, so I wrote: `if hasattr(self, "device") and self.device is not None:` everywhere, but maybe the last part is not
> > > actually needed??
> > 
> > 
> > If you implement this review comment [#647 (comment)](https://github.com/qilimanjaro-tech/qililab/pull/647#discussion_r1434872662) it will be very easy to test which if statement is the ideal, since you will only have to change one line.
> > `if hasattr(self, "device")` is correct for sure, I might also check if it is a subclass of `Device`: `isinstance(self.device, Device)`.
> 
> I'll implement your previous suggestion, so we can try it out quicker, because I am not really sure the last check is necessary!
> 
> This is the code for `CheckDeviceInitialized` we were previously using, so I guess following the same checks would be the way to go, unless someone thinks it is missing things: Screenshot 2023-12-22 at 11 16 12

After a discussion, we will go with that code, but also add the self.device is not None, for when you disconnect, since it set the device to None

File: src/qililab/instruments/mini_circuits/attenuator.py (hunk 3 above)
Comment (GuillermoAbadLopez) [reply src/qililab/instruments/mini_circuits/attenuator.py]: Removed `port_id`, changing the implementation from `system_control` as suggested!

**Comment by GuillermoAbadLopez** [comment]
Tested it with all the other PRs and the amplitude and drag are working:

Screenshot 2023-12-31 at 21 56 22

But the Intermidiate frequency gave problems, since it went through another way (bus > sys_control > instrument), take a look, and solve before merging:

Screenshot 2023-12-31 at 21 56 02

File: src/qililab/instruments/qdevil/qdevil_qdac2.py [hunk 21]
```diff
@@ -65,37 +66,45 @@ def setup(self, parameter: Parameter, value: float | str | bool, channel_id: int
         """
         self._validate_channel(channel_id=channel_id)
 
-        channel = self.device.channel(channel_id)
+        if self.is_device_initialized():
+            channel = self.device.channel(channel_id)
+        else:
+            channel = None
+
```
Comment (fedonman) [inline src/qililab/instruments/qdevil/qdevil_qdac2.py:73]: You can write this in one-liner for constistency.


```suggestion
    channel = self.device.channel(channel_id) if self.is_device_initialized() else None

```

File: src/qililab/instruments/qblox/qblox_d5a.py [hunk 22]
```diff
@@ -85,7 +84,11 @@ def setup(self, parameter: Parameter, value: float | str | bool, channel_id: int
                 f"the specified dac index:{channel_id} is out of range."
                 + " Number of dacs is 4 -> maximum channel_id should be 3."
             )
-        channel = self.dac(dac_index=channel_id)
+        if self.is_device_initialized():
+            channel = self.dac(dac_index=channel_id)
+        else:
+            channel = None
+
```
Comment (visagim) [inline src/qililab/instruments/qblox/qblox_d5a.py:91]: won't this set the channel to `None` on instrument initialization?

File: src/qililab/instruments/qblox/qblox_module.py [hunk 23]
```diff
@@ -313,9 +312,7 @@ def setup(  # pylint: disable=too-many-branches, too-many-return-statements
             return
 
         if channel_id is None:
-            if port_id is not None:
-                channel_id = self.get_sequencers_from_chip_port_id(chip_port_id=port_id)[0].identifier
-            elif self.num_sequencers == 1:
+            if self.num_sequencers == 1:
```
Comment (visagim) [inline src/qililab/instruments/qblox/qblox_module.py:315]: same issue as with the d5a, if we set a value for this, how do we re-set the value so that it is correct after instrument initialization?

File: src/qililab/instruments/qblox/qblox_s4g.py [hunk 24]
```diff
@@ -87,7 +86,11 @@ def setup(self, parameter: Parameter, value: float | str | bool, channel_id: int
                 f"the specified dac index:{channel_id} is out of range."
                 + " Number of dacs is 4 -> maximum channel_id should be 3."
             )
-        channel = self.dac(dac_index=channel_id)
+        if self.is_device_initialized():
+            channel = self.dac(dac_index=channel_id)
+        else:
+            channel = None
+
```
Comment (fedonman) [inline src/qililab/instruments/qblox/qblox_s4g.py:93]: same here

**Review by fedonman — APPROVED** [review]
Great job! 💯 Looks good to me! Left two comments on refactoring an if/else block to one-liner.

**Comment by fedonman** [comment]
You have to fulfill codecov's wishes as well though.

**Comment by GuillermoAbadLopez** [comment]
> You have to fulfill codecov's wishes as well though.

I'm not so sure about that haha, the things that are not covered are totally different and independent things, that weren't covered to start with, but are in the same files!

I might ask Albert for him to merge it, since all the changed lines are covered and it would be too much time to invest now 🙃

File: src/qililab/instruments/qdevil/qdevil_qdac2.py (hunk 21 above)
Comment (GuillermoAbadLopez) [reply src/qililab/instruments/qdevil/qdevil_qdac2.py:73]: Sure! Commiting it!

File: src/qililab/instruments/qblox/qblox_s4g.py (hunk 24 above)
Comment (GuillermoAbadLopez) [reply src/qililab/instruments/qblox/qblox_s4g.py:93]: Changed!

File: src/qililab/instruments/qblox/qblox_d5a.py (hunk 22 above)
Comment (GuillermoAbadLopez) [reply src/qililab/instruments/qblox/qblox_d5a.py]: What do you mean by instrument initialization? `initial_setup`?

**Comment by visagim** [comment]
Since @fedonman already reviewed I didn't do an extra thorough review but I did go over it and it looks fine - great job!
From my side if we get to test it on hardware and it works it's good to go (perhaps we can run autocalibration with multiple slurm tasks from https://github.com/qilimanjaro-tech/qililab-portfolio/pull/38 since it's the issue this is supposed to solve, and if everything works fine we can merge this; provided that codecov is also ok)

**Comment by GuillermoAbadLopez** [comment]
> Since @fedonman already reviewed I didn't do an extra thorough review but I did go over it and it looks fine - great job! From my side if we get to test it on hardware and it works it's good to go (perhaps we can run autocalibration with multiple slurm tasks from [qilimanjaro-tech/qililab-portfolio#38](https://github.com/qilimanjaro-tech/qililab-portfolio/pull/38) since it's the issue this is supposed to solve, and if everything works fine we can merge this; provided that codecov is also ok)

It has already been tested with auto-calibration Victor! 🙌 

Maybe we would need to do some tricky corner cases tests as well, like disconnection, setting something, and connecting, or something more complicated like that, I did test it with pytest though, but let me do it anyway 🚀 

Here are the results of some little test I did:

Screenshot 2024-01-03 at 18 41 59
Screenshot 2024-01-03 at 18 42 43
Screenshot 2024-01-03 at 18 45 05
Screenshot 2024-01-03 at 18 46 48
